src/plugins/tonegen/notification.c

```diff
@@ -0,0 +1,277 @@
+
+#define _GNU_SOURCE
+
+#include <math.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+
+#include <ngf/log.h>
+#include <trace/trace.h>
+
+#include "tonegend.h"
+#include "dbusif.h"
+#include "tone.h"
+#include "stream.h"
+#include "notification.h"
+
+#define TRACE(f, args...) trace_write(trctx, trflags, trkeys, f, ##args)
+#define LOG_CAT "tonegen-notification: "
+
+static char *notif_stream = STREAM_NOTIFICATION;
+
+struct method {
+    char  *intf;                                    /* interface name */
+    char  *memb;                                    /* method name */
+    char  *sig;                                     /* signature */
+    int  (*func)(DBusMessage *, struct tonegend *); /* implementing function */
+};
+
+static int start_notif_tone(DBusMessage *, struct tonegend *);
+static int stop_tone(DBusMessage *, struct tonegend *);
+static uint32_t linear_volume(int);
+static void notif_play(struct ausrv *ausrv, int type, uint32_t vol, int dur);
+static void notif_stop(struct ausrv *ausrv, int kill_stream);
+
+static struct method  method_defs[] = {
+    {NOTIF_INTERFACE, "StartNotificationTone", "uiu", start_notif_tone},
+    {NOTIF_INTERFACE, "StopTone", "", stop_tone},
+    {NULL, NULL, NULL, NULL}
+};
+
+static void     *notif_props = NULL;
+static uint32_t  vol_scale   = 100;
+
+int notif_init(void)
+{
+    return 0;
+}
+
+int notif_create_ngfd(struct tonegend *t)
+{
+    (void) t;
+
+    return 0;
+}
+
+int notif_create(struct tonegend *tonegend)
+{
+    struct method *m;
+    int err;
+    int sts;
+
+    for (m = method_defs, err = 0;    m->memb != NULL;    m++) {
+        sts = dbusif_register_input_method(tonegend, m->intf, m->memb,
+                                           m->sig, m->func);
+
+        if (sts < 0) {
+            N_ERROR(LOG_CAT "%s(): Can't register D-Bus method '%s'",
+                      __FUNCTION__, m->memb);
+            err = -1;
+        }
+    }
+
+    return err;
+}
+
+
+static int start_notif_tone(DBusMessage *msg, struct tonegend *tonegend)
+{
+    struct ausrv *ausrv = tonegend->ausrv_ctx;
+    uint32_t      event;
+    int32_t       dbm0;
+    uint32_t      duration;
+    uint32_t      volume;
+    int           beeptype;
+    int           success;
+
+    success = dbus_message_get_args(msg, NULL,
+                                    DBUS_TYPE_UINT32, &event,
+                                    DBUS_TYPE_INT32 , &dbm0,
+                                    DBUS_TYPE_UINT32, &duration,
+                                    DBUS_TYPE_INVALID);
+
+    if (!success) {
+        N_ERROR(LOG_CAT "%s(): Can't parse arguments", __FUNCTION__);
+        return FALSE;
+    }
+
+    volume = linear_volume(dbm0);
+
+    
+
+    TRACE("%s(): event %u  volume %d dbm0 (%u) duration %u msec",
+          __FUNCTION__, event, dbm0, volume, duration);
+
+
+    switch (event) {
+    case 66:  beeptype = TONE_DIAL;      break;
```

Comment:
Where do these magic values come from? Some of these can be ascii chars? Or better yet, have named constants for those.